File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/ApiLookup.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.apis

import slate.common.ListMap


/**
  * Container for all the registered apis that can be called dynamically.
  * This contains a lookup of api names to the actual apis.
  * e.g.
  * {
  *   "invites" => InvitesApi()
  *   "devices" => DevicesApi()
  * }
  */
class ApiLookup extends ListMap[String,ApiBase] {

}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/helpers/ApiLookup.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.helpers

import slatekit.apis.ApiReg
import slatekit.apis.ApiRegAction
import slatekit.common.*


/**
 * Provides lookup functionality to check for exposed api actions.
 * @param api: The APIReg that represents the metadata for 1 API/Class
 */
class ApiLookup(val api: ApiReg) {

    var _lookup = ListMap<String, ApiRegAction>()


    /**
     * gets a list of all the actions / methods this api supports.
     *
     * @return
     */
    fun actions(): ListMap<String, ApiRegAction> = _lookup.clone()


    /**
     * whether or not the action exists in this api
     *
     * @param action : e.g. "invite" as in "users.invite"
     * @return
     */
    fun contains(action: String): Boolean = _lookup.contains(action)


    /**
     * gets the value with the supplied key(action)
     *
     * @param action
     * @return
     */
    operator fun get(action: String): ApiRegAction? {
        return if (!contains(action))
            null
        else
            _lookup[action]!!
    }


    /**
     * gets the value with the supplied key(action)
     *
     * @param action
     * @return
     */
    fun getOpt(action: String): ApiRegAction? = _lookup[action]


    /**
     * adds a key/value to this collection
     *
     * @param action
     * @param value
     */
    fun update(action: String, value: ApiRegAction): Unit {
        _lookup = _lookup.add(action, value)
    }
}
